standards/core/rln-contract.md

```diff
@@ -0,0 +1,297 @@
+
+### Register a membership
+
+Membership registration is subject to the following conditions:
+- if there are _Expired_ memberships in the contract, the new membership MUST overwrite an expired membership;
+- the new membership SHOULD overwrite the membership that had been _Expired_ for the longest time;
+- if an _Expired_ membership A is overwritten by membership B:
+	- membership B MUST transition to _ErasedAwaitsWithdrawal_;
+	- the current total rate limit MUST be decremented by the rate limit of membership B;
+	- the contract MUST take all necessary steps to ensure that the owner of membership B can withdraw their deposit later;
+- registration MUST fail if the total rate limit of _Active_, _GracePeriod_, and _Expired_ memberships, including the one being created, would exceed the limit;
+- registration MUST fail if the requested rate limit for the new membership is lower than the minimal allowed rate limit;
+- the user MUST lock-up a deposit to register a membership;
+- the user MUST specify the rate limit of the new membership;
+- the size of the deposit MUST be calculated depending on the requested rate limit;
+- in case of successful registration:
+	- the new membership MUST be in the _Active_ state;
+	- the current total rate limit MUST be incremented by the rate limit of the new membership;
+- a newly created membership MUST have an expiration time `T` and a grace period `G` (see RECOMMENDED parameter values below).
+
+### Send a message
+
+Sending messages is handled by Waku Relay nodes, not by the RLN smart contract.
+For completeness, sending messages is mentioned here as the key Waku functionality.
+The full specification of Relay node behavior is out of scope for this document.
+
+A Relay node MUST relay a message unless:
+- the message is committed to a different epoch than the current epoch; or
+- the user has exceed their allowed rate limit for the current epoch; or
+- the RLN proof fails to prove that the message sender owns an existing membership.
+
+### Extend a membership
+
+Extending a membership is subject to the following condition:
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership owner MUST be able to extend the membership;
+- any user except the membership owner MUST NOT be able to extend the membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Owning a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
+
+### Withdraw the deposit
+
+Deposit withdrawal is subject to the following conditions:
+- the owner of a membership MUST be able to withdraw their deposit;
+- any user except the membership owner MUST NOT be able to withdraw its deposit;
+- a deposit MUST be withdrawn in full;
+- a withdrawal MUST fail if the membership is not in _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_;
```

Comment:
The owner of a membership may choose do withdraw in either of the following states: _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_. Gas costs may differ depending on the state. When withdrawing during _GracePeriod_ or _Expired_, the user assumes the gas cost of erasing the membership from the tree. In contrast, at _ErasedAwaitsWithdrawal_, the membership has already been erased from the tree (the owner of the overtaking new membership paid the cost).

There may be two issues with this state of affairs:
- there is no incentive for a user to withdraw during _GracePeriod_ even if they no longer need the membership (apart from the opportunity cost of the locked-up capital) - instead, users may choose to wait until _Expired_ and withdraw or the same gas cost;
- users don't pay the price of their expired memberships polluting the tree, increasing costs for Relay nodes (again, apart from the opportunity cost).

All in all, if the cost of locked-up capital is negligible (which it arguably is at currently recommended prices), users may only withdraw from _ErasedAwaitsWithdrawal_, and transfer the cost of tree cleanup onto new registrations. If we think this is an issue, we can think of additional incentivization to withdraw early, that is, in the ideal case, a user should withdraw (and erase their membership in the process) as soon as they no longer needs RLN functionality. An example of such incentivization would be to burn part of the deposit in proportion to the time the membership has spent in the _Expired_ or _ErasedAwaitsWithdrawal_ state.

Reply:
Let's leave this a question for future work. I don't think this incentive incompatibility, if we even can call it that, would affect much in the first spec iteration.

Reply:
Interesting concept. Maybe something to think about in terms of gas optimization for the contract cc @richard-ramos .

This may be possible to influence based on the various indexing maps used to keep track.